pil-analyzer/src/traits_resolver.rs

```diff
@@ -60,26 +91,176 @@ impl<'a> TraitsResolver<'a> {
+                    }
                 }
+            } else {
+                for impl_ in impls.iter() {
+                    let type_vars = self.trait_typevars_mapping.get(&reference.name);
+                    match type_vars {
+                        Some(type_vars) => {
+                            for t in type_vars {
+                                let tuple_args = Type::Tuple(TupleType { items: t.clone() });
+
+                                let mut unifier: Unifier = Default::default();
+
+                                let res = unifier
+                                    .unify_types(tuple_args.clone(), impl_.type_scheme.ty.clone());
+                                if res.is_ok() {
+                                    if let Some(expr) = impl_.function_by_name(trait_fn_name) {
+                                        solved_impls.push((
+                                            (reference.name.clone(), t.clone()),
+                                            Arc::clone(&expr.body),
+                                        ));
+                                    }
+                                }
+                            }
+                        }
+                        None => {
+                            continue;
+                        }
+                    }
+                }
+            }
+        }
+        solved_impls
+    }
+
+    /// Creates a dictionary that associates each child reference with a generic type along with the references that called it
+    /// From this dictionary, it resolves the parent/child relationships to obtain the value of the generic type
+    fn build_reference_path(
+        definitions: &Vec<(&String, &Expression)>,
+        defined_traits: &HashSet<&String>,
+    ) -> HashMap<String, HashSet<Vec<Type>>> {
+        let mut result = HashMap::new();
+
+        // Compute all pairs of references (name, type) for each child and store them associated with the parent's name
+        // ["Child": [
+        //      ("Parent1", [Some(type_args)]),
+        //      ("Parent2", [Some(type_args)])
+        // ],...]
+        for (symbol, expr) in definitions {
+            let mut references = Vec::new();
+            expr.all_children().for_each(|e| {
+                if let Expression::Reference(
+                    _,
+                    Reference::Poly(PolynomialReference {
+                        name,
+                        type_args: Some(type_args),
+                        ..
+                    }),
+                ) = e
+                {
+                    if (!type_args.is_empty()) && (name != *symbol) {
+                        references.push((name, type_args));
+                    }
+                }
+            });
+            for (name, type_args) in references {
+                result
+                    .entry(name)
+                    .or_insert_with(Vec::new)
+                    .push((*symbol, type_args));
+            }
```

Comment:
Do we need to collect in `references` first? Why not insert into `result` directly?